tests/nn/data_parallel/test_fsdp_freezing_weights.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-class-docstring
+# pylint: disable=missing-function-docstring
+
+""" Test FSDP with some params frozen. """
+
+
+import tempfile
+
+import pytest
+import torch
+import torch.multiprocessing as mp
+import torch.nn as nn
+from torch.nn.parallel import DistributedDataParallel
+import torch.optim as optim
+
+from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
+from fairscale.utils.testing import dist_init, objects_are_equal, rmf, skip_if_single_gpu, teardown
+
+
+class Model(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.trunk = nn.Sequential(
+            nn.Conv2d(3, 64, kernel_size=3),
+            nn.ReLU(inplace=True),
+            nn.AdaptiveAvgPool2d(output_size=(1, 1)),
+            nn.Flatten(),
+        )
+        self.head = nn.Linear(64, 10)
+
+    def forward(self, x):
+        return self.head(self.trunk(x))
+
+
+def _create_model(with_fsdp: bool):
+    model = Model()
+    if with_fsdp:
+        model.trunk = FSDP(model.trunk)
+        model.head = FSDP(model.head)
+    return model
+
+
+def _distributed_worker(
+    gpu_id: int, world_size: int, with_fsdp: bool, tempfile_name, unused, rank_0_output, expected_state
+):
+    torch.cuda.set_device(gpu_id)
+
+    rank = gpu_id
+    result = dist_init(rank, world_size, tempfile_name, unused)
+    assert result, "Dist init failed"
+
+    torch.manual_seed(0)
+    torch.backends.cudnn.deterministic = True
+    batch = torch.randn(size=(2, 3, 224, 224)).cuda()
+
+    model = _create_model(with_fsdp)
+    model = model.cuda()
+
+    # freezing the trunk.
+    for name, param in model.named_parameters():
+        if "trunk" in name:
+            param.requires_grad = False
```

Comment:
From what I understand, VISSL is doing something different, the grads are removed after the BW pass. requires_grad being changed makes more sense to me since you actually save time and space during the BW instead of computing then wiping it. This being said DDP does not support changing .requires_grad over time, the model needs a re-wrap in that case, may be the reason why this was done like that to begin with

Reply:
Thanks for taking a look Ben and great insights!!!

Indeed, setting grad to None is for a different type of freezing used in vissl. And we have tested that in vissl and seems to work fine. This is for "linear evaluation" type of training, where only the head is allowed to learn. I.e. vissl uses both methods.

In FSDP, change requires_grad seems to be fine but I haven't tested all cases exhaustively. It might not work with different flattening and nested wrap boundaries, etc.